compiler/hash-reporting/src/render.rs

```diff
@@ -0,0 +1,464 @@
+//! Hash diagnostic report rendering logic. Within this renderer,
+//! there are currently two modes: `line` and `block`. Within this
+//! module, the modes are often referred to either being line or block.
+//! The difference between the two modes is how they are displayed
+//! when rendered. The `line` mode is designed to highlight spans that
+//! are contained to a single line, whereas `block` mode is designed to
+//! for bulkier spans that might take up many lines. More specific examples
+//! of both modes are located in the documentation of the functions
+//! that are responsible for rendering either mode.
+
+use std::{
+    fmt,
+    iter::{once, repeat, Enumerate, Skip, Take},
+    str::Split,
+};
+
+use hash_source::{location::SourceLocation, SourceMap};
+use hash_utils::path::adjust_canonicalization;
+
+use crate::{
+    highlight::{highlight, Colour, Modifier},
+    report::{ReportCodeBlock, ReportCodeBlockInfo, ReportElement, ReportKind, ReportNote},
+};
+
+/// Character used to denote the span of the diagnostic for the `line` view.
+const LINE_DIAGNOSTIC_MARKER: char = '^';
+
+/// Character used to denote the span of the diagnostic for the `block` view.
+const BLOCK_DIAGNOSTIC_MARKER: char = '-';
+
+/// Character used to connect block views
+const DIAGNOSTIC_CONNECTING_CHAR: &str = "|";
+
+/// Function to compute a row and column number from a given source string
+/// and an offset within the source. This will take into account the number
+/// of encountered newlines and characters per line in order to compute
+/// precise row and column numbers of the span.
+pub(crate) fn offset_col_row(
+    offset: usize,
+    source: &str,
+    non_inclusive: bool,
+) -> (/* col: */ usize, /* row: */ usize) {
```

Comment:
Maybe it might be worth it to make a `ColRowOffset` struct with fields `col` and `row`?